docs/CONTROLFLOW.md

```diff
@@ -0,0 +1,319 @@
+# Control Flow Obfuscation
+
+> **Feature is experimental**. To activate it, set environment variable `GARBLE_EXPERIMENTAL_CONTROLFLOW=1`
+
+### Mechanism
+
+
+Control flow obfuscation works in several stages:
+
+1) Collect functions with `//garble:controlflow` comment
+2) Converts [go/ast](https://pkg.go.dev/go/ast) representation to [go/ssa](https://pkg.go.dev/golang.org/x/tools/go/ssa)
+3) Applies [block splitting](#block-splitting)
+4) Generates [junk jumps](#junk-jumps)
+5) Applies [control flow flattening](#control-flow-flattening)
+6) Converts go/ssa back into go/ast
+
+### Example usage
+
+```go
+//garble:controlflow flatten_passes=1 junk_jumps=1 block_splits=1
+func main() {
+	println("Hellow world!")
+}
+```
+
+### Parameter explanation
+
+> Unlike other garble features (which just work), we recommend that you understand how parameters affect control flow obfuscation and what caveats exists.
+
+> Code snippets below without name obfuscation for better readability.
+
+#### Block splitting
+
+Param: `block_splits` (default: `0`)
+
+> Warning: this param affects resulting binary only when used in combination with [flattening](#control-flow-flattening)
+
+
+Block splitting `block_splits` times splits largest block into 2 parts of random size. If there is no suitable block (number of ssa instructions in block is less than 3) work stops without errors.
+
+This param is very useful if your code has few branches (`if`, `switch` etc.).
+
+Input:
+```go
+package main
+
+// Note that the block_splits value is very large, so code blocks are splitted into the smallest possible blocks.
+//garble:controlflow flatten_passes=0 junk_jumps=0 block_splits=1024
+func main() {
+	println("1")
+	println("2")
+	println("3")
+	println("4")
+	println("5")
+}
+```
+
+Result:
+
+```go
+func main() {
+	{
+		println("1")
+		goto _s2a_l3
+	}
+_s2a_l1:
+	{
+		println("3")
+		goto _s2a_l4
+	}
+_s2a_l2:
+	{
+		println("5")
+		return
+	}
+_s2a_l3:
+	{
+		println("2")
+		goto _s2a_l1
+	}
+_s2a_l4:
+	{
+		println("4")
+		goto _s2a_l2
+	}
+}
+ 
+```
+
+
+#### Junk jumps
+
+Param: `junk_jumps` (default: `0`)
+
+> Warning: this param affects resulting binary only when used in combination with [flattening](#control-flow-flattening)
+
+Junk jumps adds `junk_jumps` times to random blocks junk jump, can make a chain of jumps. This function is useful for linearly increasing function complexity.
```

Comment:
Junk jumps adds jumps to random blocks `junk_jumps` times, these inserted jumps can also form a chain. This param is useful for linearly increasing the functions complexity.